lib/seattleflu/db/cli/command/etl/manifest.py

```diff
@@ -233,11 +357,142 @@ def upsert_sample(db: DatabaseSession,
+    if not identifier:
+        LOG.warning(f"Skipping record with unknown barcode «{barcode}»")
+        mark_skipped(db, record.id)
+        return None
+    
+    assert identifier.set_name in identifier_set, \
+        (f"Identifier found in set «{identifier.set_name}», " + 
+        f"not {identifier_set}")
+
+    return identifier
+
+
+def utm_sample_details(document: dict) -> dict:
+    """
+    Parse out UTM sample details from a kit manifest record *document* 
+
+    TODO: Accept utm_details from option/config
+    """
+    utm_details = {"aliquot_date", "aliquots", "racks"}
+    additional_details = {}
+    for column in utm_details:
+        new_details = document.pop(column, None)
+        if new_details and any(new_details):
+            additional_details.update({ column: new_details })
+    
+    return additional_details
+    
+
+def update_test_strip(db: DatabaseSession, document: dict): 
+    """
+    Find identifier that matches the test_strip barcode within *document*.
+    Updates *document* to have both the test_strip barcode and 
+    the identifier if found.
+    """
+    strip_barcode = document["test_strip"]
+    strip_identifier = find_identifier(db, strip_barcode)
+
+    if (strip_identifier and 
+            strip_identifier.set_name == "test-strips-fluathome.org"):
+        document["test_strip"] = {
+            "uuid": strip_identifier.uuid,
+            "barcode": strip_identifier.barcode
+        }
+    else:
+        document["test_strip"] = {
+            "uuid": None,
+            "barcode": strip_barcode
+        }
+        
+
+def upsert_kit(db: DatabaseSession, 
+               kit_identifier: str,
+               rdt_sample_id: int,
+               utm_sample_id: int,
+               additional_details: dict) -> Any:
+    """
+    Upsert kit by its *identifier*
+
+    An existing kit has its *rdt_sample_id* and *utm_sample_id* updated, 
+    and the provided *additional_details* are merged (at the top-level only) 
+    into the existing kit details, if any.
+    """
+    LOG.debug(f"Upserting kit «{kit_identifier}»")
+
+    data = {
+        "identifier": kit_identifier,
+        "rdt_sample_id": rdt_sample_id,
+        "utm_sample_id": utm_sample_id,
+        "additional_details": Json(additional_details)
+    }
+
+    kit = db.fetch_row("""
+        insert into warehouse.kit (identifier, rdt_sample_id, utm_sample_id, details)
+            values (%(identifier)s, 
+                    %(rdt_sample_id)s,
+                    %(utm_sample_id)s, 
+                    %(additional_details)s)
+        
+        on conflict (identifier) do update
+            set rdt_sample_id = %(rdt_sample_id)s,
+                utm_sample_id = %(utm_sample_id)s,
+                details = coalesce(kit.details, '{}') || %(additional_details)s
+        
+        returning kit_id as id, identifier, encounter_id, rdt_sample_id, utm_sample_id,
+            case when xmax::text::int = 0 then 'created' else 'updated' end as status
```

Comment:
This was just an answer I found on [stackoverflow](https://stackoverflow.com/questions/34762732/how-to-find-out-if-an-upsert-was-an-update-with-postgresql-9-5-upsert) on how to determine if an update happened during as upsert. But it may be no longer required because I was thinking of separating things out like you did in `upsert_sample` to check if we are moving around kit-encounter references